airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonToAvroSchemaConverter.java

```diff
@@ -200,17 +200,22 @@ Schema getSingleFieldType(final String fieldName, final JsonSchemaType fieldType
     return fieldSchema;
   }
 
-  List<Schema> getSchemasFromTypes(final String fieldName, final ArrayNode types) {
-    return MoreIterators.toList(types.elements())
-        .stream()
-        .flatMap(definition -> getNonNullTypes(fieldName, definition).stream().flatMap(type -> {
-          final Schema singleFieldSchema = getSingleFieldType(fieldName, type, definition);
-          if (singleFieldSchema.isUnion()) {
-            return singleFieldSchema.getTypes().stream();
-          } else {
-            return Stream.of(singleFieldSchema);
-          }
-        }))
+  List<Schema> getSchemasFromTypes(final String fieldName, final ArrayNode types, boolean combined) {
+    return StreamUtils
+        .zipWithIndex(MoreIterators.toList(types.elements())
+            .stream())
+        .flatMap(definition -> getNonNullTypes(fieldName, definition.getValue()).stream()
+            .flatMap(type -> {
+              final Schema singleFieldSchema = getSingleFieldType(combined
+                  // avro schema allows only unique names of records in case of union
+                  ? fieldName.concat("_").concat(String.valueOf(definition.getIndex()))
```

Comment:
Under which circumstance will the names have conflict here? The stream calls `distinct` at the end, so any redundant types will be removed.

Is the change here related to `$ref`? Currently the `COMBINED` type is only used for processing combined restrictions, i.e. `oneOf`, `allOf`, and `anyOf`. I don't think it can or should be be used to handle `$ref`.

Reply:
> Under which circumstance will the names have conflict here? The stream calls `distinct` at the end, so any redundant types will be removed.
> 
> Is the change here related to `$ref`? Currently the `COMBINED` type is only used for processing combined restrictions, i.e. `oneOf`, `allOf`, and `anyOf`. I don't think it can or should be be used to handle `$ref`.

i find this behaviour when converting objects from this schema (Mailchimp)
```
{
  "type": "array",
  "title": "Segment Type",
  "description": "Segment match conditions. There are multiple possible types, see the [condition types documentation](https://mailchimp.com/developer/marketing/docs/alternative-schemas/#segment-condition-schemas).",
  "items": {
    "x-discriminator": {
      "type": "string",
      "propertyName": "condition_type"
    },
    "oneOf": [
      {
        "$ref": "aimSegment.json"
      },
      {
        "$ref": "automationSegment.json"
      },
      {
        "$ref": "campaignPollSegment.json"
      },
      {
        "$ref": "conversationSegment.json"
      },
      {
        "$ref": "dateSegment.json"
      },
      {
        "$ref": "emailClientSegment.json"
      },....
```
distinct call will not help in this case, it just decreases count of avro records in union, but anyway all records will have name "conditions.items" and that is not allowed in avro union

Reply:
I am not sure what the `conditions.items` is. It's not in any of the Mailchimp modified in this PR.

Following the above example for Mailchimp, what would be the final schema after running it through the converter?

Reply:
@tuliren sorry for delay. This should be resolved in scope of https://github.com/airbytehq/airbyte/pull/8574 so i will revert these changes from this PR and leave only updated schemas of Mailchimp

Reply:
@tuliren reverted changes from JsonToAvroSchemaConverter. Please advice if anything else should be implemented in scope of this task